quodlibet/quodlibet/ext/events/libraryfileorganizing.py

```diff
@@ -0,0 +1,119 @@
+from quodlibet.util.library import scan_library
+
+
+class MyPlugin(EventPlugin, PluginConfigMixin):
+    PLUGIN_ID = "libraryfileorganizing"
+    PLUGIN_NAME = _("Automatically Update Library Files")
+    PLUGIN_DESC = _("After making any update (namely, to tags), update "
+                    "the file to a new location/name based on its tags in "
+                    "a form that you specify. This is most useful for "
+                    "automatically keeping organized a central music "
+                    "directory.")
+
+    DEFAULT_PAT = "~/Music/<albumartist|<albumartist>|" + \
+                  "<artist|<artist>|Unknown Artist>>/" + \
+                  "<album|<album>|Unknown Album>/" + \
+                  "<albumartist|<albumartist>|" + \
+                  "<artist|<artist>|Unknown Artist>> " + \
+                  "(<album|<album>|Unknown Album>) - " + \
+                  "<~#track|<~#track><~#disc| (<~#disc>)> - >" + \
+                  "<title|<title>|Unknown Title>>"
+    CFG_PAT_PLAYING = "naming_pattern"
+
+    def plugin_on_changed(self, songs):
+        song = songs[0]
+        current_location = song("~filename")
+        file_split = os.path.splitext(current_location)
+        file_extension = file_split[1]
+        psbl_new_location = os.path.abspath(
+            os.path.expanduser((Pattern(self.config_get(self.CFG_PAT_PLAYING,
+                                                        self.DEFAULT_PAT)) %
+                                song) + file_extension))
+
+        if(not(psbl_new_location == re.sub(" \\([0-9]+\\)$",
+                                           "",
+                                           file_split[0]) + file_extension)):
+            dirname = os.path.dirname(psbl_new_location)
+            if(not(os.path.exists(dirname))):
+                os.makedirs(dirname)
+
+            if(os.path.isfile(psbl_new_location)):
+                num_to_use = 1
+                start_with = os.path.splitext(
+                    os.path.basename(psbl_new_location))[0]
+                length_chk = len(start_with) + 2
+
+                for f in sorted(os.listdir(dirname)):
+                    if(f.startswith(start_with)):
+                        temporary = f[length_chk:]
+
+                        if(temporary[:temporary.find(")")] == str(num_to_use)):
+                            num_to_use += 1
+
+                psbl_new_location = dirname + "/" + start_with + \
+                                    " (" + str(num_to_use) + ")" + \
+                                    file_extension
+
+            shutil.move(current_location, os.path.normpath(psbl_new_location))
+
+            self.removeEmptyDirs(os.path.dirname(current_location))
+
+            scan_library(app.library, False)
+
+    def PluginPreferences(self, parent):
+        outer_vb = Gtk.VBox(spacing=12)
+        vb = Gtk.VBox(spacing=12)
+
+        # Naming Pattern
+        hb = Gtk.HBox(spacing=6)
+        entry = UndoEntry()
```

Comment:
The pattern is a bit awkward to edit since its so long. Consider using a PatternEditBox instance instead.